### ISSUE TITLE
Register a pluggable component in k8s failed

### ISSUE THREAD

**Issue by Bryce-huang** [dapr/dapr#5858]
## In what area(s)?

/area sidecar

## What version of Dapr?

1.9.5
> edge: output of `git describe --dirty`

## Expected Behavior

 
pluggable component registered successfully or report the error

## Actual Behavior

look like darpd time out
```
time="2023-01-31T08:20:28.501003333Z" level=info msg="starting Dapr Runtime -- version 1.9.5 -- commit f5f847eef8721d85f115729ee9efa820fe7c4cd3" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501039417Z" level=info msg="log level set to: debug" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501192792Z" level=info msg="metrics server started on :9090/" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.metrics type=log ver=1.9.5
time="2023-01-31T08:20:28.501347292Z" level=info msg="loading default configuration" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501360167Z" level=debug msg="Resiliency is not enabled." app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501383167Z" level=info msg="kubernetes mode configured" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501389208Z" level=info msg="app id: dapr-queue" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501567958Z" level=info msg="mTLS enabled. creating sidecar authenticator" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.501750333Z" level=info msg="trust anchors and cert chain extracted successfully" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime.security type=log ver=1.9.5
time="2023-01-31T08:20:28.501767417Z" level=info msg="authenticator created" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.507873458Z" level=info msg="Dapr trace sampler initialized: DaprTraceSampler(P=0.000100)" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.507917167Z" level=info msg="Initialized name resolution to kubernetes" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=dapr.runtime type=log ver=1.9.5
time="2023-01-31T08:20:28.507938042Z" level=debug msg="loading pluggable components under path /tmp/dapr-components-sockets" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=pluggable-components-grpc-connector type=log ver=1.9.5
time="2023-01-31T08:20:28.507985Z" level=debug msg="using socket defined at 'unix:///tmp/dapr-components-sockets/queue.sock'" app_id=dapr-queue instance=dapr-queue-6b7648d889-zm7gh scope=pluggable-components-grpc-connector type=log ver=1.9.5
```

## Steps to Reproduce the Problem

1. make the pluggable compponet deploy eg:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dapr-queue
  labels:
    app: dapr-queue
spec:
  replicas: 1
  selector:
    matchLabels:
      app: dapr-queue
  template:
    metadata:
      labels:
        app: dapr-queue
      annotations:
        dapr.io/unix-domain-socket-path: "/tmp/dapr-components-sockets" 
        dapr.io/app-id: "dapr-queue"
        dapr.io/enabled: "true"
        dapr.io/log-level: "debug"
        dapr.io/sidecar-liveness-probe-delay-seconds: "45"
        dapr.io/sidecar-readiness-probe-delay-seconds: "45"
        dapr.io/sidecar-liveness-probe-threshold: "5"
        dapr.io/sidecar-readiness-probe-threshold: "5"
    spec:
      volumes: 
        - name: dapr-unix-domain-socket
          emptyDir: {}
      containers:
        - name: component
          volumeMounts: 
            - name: dapr-unix-domain-socket
              mountPath: /tmp/dapr-components-sockets
          image: brycehuang/pubsub:v1
          imagePullPolicy: Always 
          lifecycle:
            postStart:
              exec: 
                command: ["/bin/sh","-c",'sleep 30']
          env:
           - name: TZ
             value: Asia/Shanghai
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: queue
scopes:
  - publish
  - subscribe
spec:
  type: pubsub.queue
  version: v1
  metadata: [] 
```

## Release Note

RELEASE NOTE:

**Comment by mcandeia** [comment]
Do you have any code I can take a look at? If not, have you set the socket permissions on the pluggable component side? When a process create an UDS it should allow other processes to RW on the same socket, so you might need to set those permissions.

Go example: https://github.com/dapr-sandbox/components-go-sdk/blob/main/socket_unix.go#L22
.NET Example: https://github.com/artursouza/kubecon-na-2022/blob/main/pluggable-components/memstore/Program.cs#L62
Java example: https://github.com/artursouza/kubecon-na-2022/blob/main/pluggable-components/discord-binding/src/main/java/binding/Server.java#L70

**Comment by Bryce-huang** [comment]
> 
 The sample code from here:
https://github.com/tmacam/DaprPluggableComponent-Java

**Comment by Bryce-huang** [comment]
It would be better for daprd to timeout rather than block when connecting to the sock file, so that people can easily diagnose the problem

**Comment by mcandeia** [comment]
Things has changed a lot since the code you provided, use the one I provided https://github.com/artursouza/kubecon-na-2022/blob/main/pluggable-components/discord-binding/src/main/java/binding/Server.java#L70 instead, and let me know if everything is working fine.

> It would be better for daprd to timeout rather than block when connecting to the sock file, so that people can easily diagnose the problem
Yeah, actually it should

**Comment by mcandeia** [comment]
any updates @Bryce-huang ?

**Comment by Bryce-huang** [comment]
> any updates @Bryce-huang ?

Thanks for your help, I am trying to code the pubsub component, but when I implement the pullMessages method, I have a doubt, should I get which topics are subscribed directly through dapr sdk? This doesn't seem to be available through method arguments.

**Comment by Bryce-huang** [comment]
I wrote a new code based on your suggestion, and updated to k8s again. The log of daprd shows that the pluggable component was registered successfully.
```
time="2023-02-01T12:43:03.177723632Z" level=info msg="app id: queue" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:43:03.177877174Z" level=info msg="mTLS enabled. creating sidecar authenticator" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:43:03.177986174Z" level=info msg="trust anchors and cert chain extracted successfully" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime.security type=log ver=1.9.5
time="2023-02-01T12:43:03.178002965Z" level=info msg="authenticator created" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:43:03.18281484Z" level=info msg="Dapr trace sampler initialized: DaprTraceSampler(P=0.000100)" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:43:03.182851549Z" level=info msg="Initialized name resolution to kubernetes" app_id=queue instance=queue-7f995468bf-txtv6 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:43:03.182874257Z" level=debug msg="loading pluggable components under path /tmp/dapr-components-sockets" app_id=queue instance=queue-7f995468bf-txtv6 scope=pluggable-components-grpc-connector type=log ver=1.9.5
time="2023-02-01T12:43:03.18292034Z" level=debug msg="using socket defined at 'unix:///tmp/dapr-components-sockets/queue.sock'" app_id=queue instance=queue-7f995468bf-txtv6 scope=pluggable-components-grpc-connector type=log ver=1.9.5
time="2023-02-01T12:43:04.154588091Z" level=debug msg="found 1 pluggable component services" app_id=queue instance=queue-7f995468bf-txtv6 scope=pluggable-components-grpc-connector type=log ver=1.9.5
time="2023-02-01T12:43:04.164290549Z" level=info msg="pluggable component 'queue' was successfully registered for 'dapr.proto.components.v1.PubSub'" app_id=queue instance=queue-7f995468bf-txtv6 scope=pluggable-components-grpc-connector type=log ver=1.9.5
```
I thought it was successful, but when I added the consumer and producer, the dpard's log of them shows 

```
time="2023-02-01T12:58:27.770133754Z" level=info msg="Stopping PubSub subscribers and input bindings" app_id=consumer instance=consumer-c4567c48-b7fbv scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:58:27.770138338Z" level=info msg="Stopping Dapr APIs" app_id=consumer instance=consumer-c4567c48-b7fbv scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:58:27.770155588Z" level=info msg="Waiting 5s to finish outstanding operations" app_id=consumer instance=consumer-c4567c48-b7fbv scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:58:32.771012048Z" level=info msg="Shutting down all remaining components" app_id=consumer instance=consumer-c4567c48-b7fbv scope=dapr.runtime type=log ver=1.9.5
time="2023-02-01T12:58:32.771127757Z" level=fatal msg="process component queue error: [CREATE_COMPONENT_FAILURE]: initialization error occurred for queue (pubsub.queue/v1): couldn't find message bus pubsub.queue/v1" app_id=consumer instance=consumer-c4567c48-b7fbv scope=dapr.runtime type=log ver=1.9.5

```
The sample code is : https://github.com/Bryce-huang/dapr-pubsub-example

**Comment by Bryce-huang** [comment]
@mcandeia 
The pubsub pluggable component of dapr does not have too many reference examples, which is too difficult for those who are not familiar with the source code of dapr.
So I would like to ask your help to check what is wrong, thank you very much!

**Comment by mcandeia** [comment]
Oh, I see @Bryce-huang. You are the second person that made the same mistake, but that's an issue with our documentation not your fault, I think we are not clear enough about that.

The problem is because the pluggable component itself should be deployed alongside your containers within the same pod, not in a separated pod as you did,

try this:

producer:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: producer
  labels:
    app: producer
spec:
  replicas: 1
  selector:
    matchLabels:
      app: producer
  template:
    metadata:
      labels:
        app: producer
      annotations:
        dapr.io/unix-domain-socket-path: "/tmp/dapr-components-sockets"
        dapr.io/app-id: "producer"
        dapr.io/enabled: "true"
        dapr.io/log-level: "debug"
    spec:
      volumes:
        - name: dapr-unix-domain-socket
          emptyDir: {}
      containers:
        - name: producer 
          image: brycehuang/producer:v1
          imagePullPolicy: Always
          env:
            - name: TZ
              value: Asia/Shanghai
        - name: component
          volumeMounts:
            - name: dapr-unix-domain-socket
              mountPath: /tmp/dapr-components-sockets
          image: brycehuang/pubsub:v3
          imagePullPolicy: Always
          lifecycle:
            postStart:
              exec:
                command: ["/bin/sh","-c","sleep 30"]
          env:
            - name: TZ
              value: Asia/Shanghai
```

the consumer should follow the same idea, you don't need the `pluggable-deploy.yaml`

I'm not sure about the implementation itself but I'm assuming that you have tested it locally, right?

**Comment by Bryce-huang** [comment]
@mcandeia Thanks for your reply, after I use the new yaml, daprd still doesn't work. It shows pluggable component loaded in the daprd log and I found in the log of my pluggable component that daprd has called my init and ping methods. it's making me helpless.

The new Yaml :

```yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: consumer
  labels:
    app: consumer
spec:
  replicas: 1
  selector:
    matchLabels:
      app: consumer
  template:
    metadata:
      labels:
        app: consumer
      annotations:
        dapr.io/app-id: "consumer"
        dapr.io/enabled: "true"
        dapr.io/log-level: "debug"
        dapr.io/app-port: "8081"
        dapr.io/unix-domain-socket-path: "/tmp/dapr-components-sockets"
        dapr.io/sidecar-liveness-probe-delay-seconds: "45"
        dapr.io/sidecar-readiness-probe-delay-seconds: "45"
        dapr.io/sidecar-liveness-probe-threshold: "5"
        dapr.io/sidecar-readiness-probe-threshold: "5"
    spec:
      volumes:
        - name: dapr-unix-domain-socket
          emptyDir: {}
      containers:
        - name: component
          volumeMounts:
            - name: dapr-unix-domain-socket
              mountPath: /tmp/dapr-components-sockets
          image: brycehuang/pubsub:v3
          imagePullPolicy: Always
          lifecycle:
            postStart:
              exec:
                command: [ "/bin/sh","-c","sleep 30" ]
          env:
            - name: TZ
              value: Asia/Shanghai
        - name: consumer
          image: brycehuang/consumer:v1
          imagePullPolicy: Always
          env:
            - name: TZ
              value: Asia/Shanghai
```
https://github.com/Bryce-huang/dapr-pubsub-example/blob/main/config/consumer.yaml

The logs of daprd

```
time="2023-02-02T11:03:44.406281596Z" level=debug msg="loading component. name: kubernetes, type: secretstores.kubernetes/v1" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.406309346Z" level=info msg="loading components" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.407369262Z" level=info msg="component loaded. name: kubernetes, type: secretstores.kubernetes/v1" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.407504221Z" level=debug msg="found component. name: queue, type: pubsub.queue/v1" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.407533679Z" level=info msg="waiting for all outstanding components to be processed" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.407574929Z" level=debug msg="loading component. name: queue, type: pubsub.queue/v1" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.407632554Z" level=debug msg="using socket defined at 'unix:///tmp/dapr-components-sockets/queue.sock'" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=pluggable-components-grpc-connector type=log ver=1.9.5
time="2023-02-02T11:03:44.478008012Z" level=info msg="component loaded. name: queue, type: pubsub.queue/v1" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.478058762Z" level=info msg="all outstanding components processed" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.478070679Z" level=info msg="gRPC proxy enabled" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.478235346Z" level=info msg="enabled gRPC tracing middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.api type=log ver=1.9.5
time="2023-02-02T11:03:44.478280262Z" level=info msg="enabled gRPC metrics middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.api type=log ver=1.9.5
time="2023-02-02T11:03:44.478362721Z" level=info msg="enabled gRPC tracing middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.api type=log ver=1.9.5
time="2023-02-02T11:03:44.478371762Z" level=info msg="enabled gRPC metrics middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.api type=log ver=1.9.5
time="2023-02-02T11:03:44.478407262Z" level=info msg="API gRPC server is running on port 50001" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.479151012Z" level=info msg="enabled metrics http middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:03:44.479278762Z" level=info msg="enabled tracing http middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:03:44.479435929Z" level=info msg="enabled metrics http middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:03:44.479556429Z" level=info msg="enabled tracing http middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:03:44.479581637Z" level=info msg="http server is running on port 3500" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.479588096Z" level=info msg="The request body size parameter is: 4" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.479618762Z" level=info msg="enabled gRPC tracing middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.internal type=log ver=1.9.5
time="2023-02-02T11:03:44.479631637Z" level=info msg="enabled gRPC metrics middleware" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.internal type=log ver=1.9.5
time="2023-02-02T11:03:44.479646512Z" level=info msg="sending workload csr request to sentry" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.internal type=log ver=1.9.5
time="2023-02-02T11:03:44.493711346Z" level=info msg="certificate signed successfully" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.internal type=log ver=1.9.5
time="2023-02-02T11:03:44.493839887Z" level=info msg="internal gRPC server is running on port 50002" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.493853679Z" level=info msg="application protocol: http. waiting on port 8081.  This will block until the app is listening on that port." app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime type=log ver=1.9.5
time="2023-02-02T11:03:44.493963262Z" level=info msg="starting workload cert expiry watcher. current cert expires on: 2023-02-03 11:18:44 +0000 UTC" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.grpc.internal type=log ver=1.9.5
time="2023-02-02T11:04:32.157321548Z" level=debug msg="{ERR_HEALTH_NOT_READY dapr is not ready}" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:04:32.15732234Z" level=debug msg="{ERR_HEALTH_NOT_READY dapr is not ready}" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:04:38.156656218Z" level=debug msg="{ERR_HEALTH_NOT_READY dapr is not ready}" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5
time="2023-02-02T11:04:38.156778926Z" level=debug msg="{ERR_HEALTH_NOT_READY dapr is not ready}" app_id=consumer instance=consumer-6855bfdd56-r2kk4 scope=dapr.runtime.http type=log ver=1.9.5

```

**Comment by Bryce-huang** [comment]
The grpc service of my pluggable component works well.
I've tested all the methods defined in `pubsub.proto` but it doesn't work with daprd, am I missing something? my sample is here: https://github.com/Bryce-huang/dapr-pubsub-example/tree/main/pubsub

**Comment by mcandeia** [comment]
>  but it doesn't work with daprd
What do you mean by "doesn't work", did your app call dapr and it is receiving an error, what is the error? I'm not seeing any significant error in the logs you pasted aside from "ERR_HEALTH_NOT_READY" which I think it is not related to pluggable components itself.

Also, as I said could you try to run it locally first using a docker-compose file like [this](https://github.com/dapr-sandbox/components-go-sdk/blob/main/examples/docker-compose.yml) ?

Or even using dapr cli if you prefer.

**Comment by Bryce-huang** [comment]
> > but it doesn't work with daprd
> > What do you mean by "doesn't work", did your app call dapr and it is receiving an error, what is the error? I'm not seeing any significant error in the logs you pasted aside from "ERR_HEALTH_NOT_READY" which I think it is not related to pluggable components itself.
> 
> Also, as I said could you try to run it locally first using a docker-compose file like [this](https://github.com/dapr-sandbox/components-go-sdk/blob/main/examples/docker-compose.yml) ?
> 
> Or even using dapr cli if you prefer.

Oh, I mean daprd doesn't work, but I can't diagnose what is causing daprd not to work through daprd's logs.

**Comment by mcandeia** [comment]
is that fixed @Bryce-huang ? 

if you can, please let us know what was causing the problem and what was the actual fix? Thanks

**Comment by Bryce-huang** [comment]
> is that fixed @Bryce-huang ? 
> 
> 
> 
> if you can, please let us know what was causing the problem and what was the actual fix? Thanks

Sorry for the late reply.  I found in the debug that the app port configuration is wrong, which makes daprd unable to work.  In addition, I think the debug log of daprd should increase the prompt of error message.  Finally, I found that my pluggable component can send messages to daprd, and daprd can also receive messages.  But daprd can't send the message to the consumer.  But I think this is another issue, so I close this issue, thank you very much for your help, hope you have a nice weekend.
